### PR TITLE
DiscIO: Use std::optional in Volume and Blob

### PR DESCRIPTION
**Pull request by JosJuice**
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/5530)

### CODE REVIEW

File: Source/Core/DiscIO/Volume.cpp [hunk 1]
```diff
@@ -93,26 +94,25 @@ std::unique_ptr<IVolume> CreateVolumeFromFilename(const std::string& filename)
...
-    return std::make_unique<CVolumeGC>(std::move(reader));
-  if (wii_magic == 0x5D1C9EA3 && wii_container_magic == 0)
-    return std::make_unique<CVolumeWiiCrypted>(std::move(reader));
+  const std::optional<u32> wii_magic = reader->ReadSwapped<u32>(0x18);
+  if (wii_magic == u32(0x5D1C9EA3))
+  {
+    const std::optional<u32> wii_container_magic = reader->ReadSwapped<u32>(0x60);
+    if (wii_container_magic == u32(0))
+      return std::make_unique<CVolumeWiiCrypted>(std::move(reader));
+    else
```
Comment (leoetlino) [inline Source/Core/DiscIO/Volume.cpp]: Unnecessary else

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by leoetlino** [comment]
Reviewed 23 of 23 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Core/Core/ConfigManager.cpp, line 723 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/5530#-Klo9-8MRf9NysAKSkns:-Klo9-8MRf9NysAKSknt:b1y3ca0) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d28c253e74c37e4f506f7d5c551e7c21b24401b4/Source/Core/Core/ConfigManager.cpp#L723)):*
> ```cpp
>                                      const DiscIO::Partition& partition)
> {
>   if (const std::optional<u16> revision = volume.GetRevision(partition))
> ```

Would it be better to just use value_or(0) instead of requiring a valid revision?

---

*[Source/Core/DiscIO/Volume.cpp, line 98 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/5530#-Klo8c9y6YjxSUtif6TB:-Klo8c9y6YjxSUtif6TC:buy7rgg) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d28c253e74c37e4f506f7d5c551e7c21b24401b4/Source/Core/DiscIO/Volume.cpp#L98)):*
> ```cpp
>   // Check for Wii
>   const std::optional<u32> wii_magic = reader->ReadSwapped<u32>(0x18);
>   if (wii_magic == u32(0x5D1C9EA3))
> ```

Unrelated to this PR, but it'd probably be worth extracting these magic values to named constants in a future PR.

---

*[Source/Core/DiscIO/VolumeDirectory.h, line 54 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/5530#-Klo8u90c3-ANZHawGMM:-Klo8u90c3-ANZHawGMN:bgjnpmt) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d28c253e74c37e4f506f7d5c551e7c21b24401b4/Source/Core/DiscIO/VolumeDirectory.h#L54)):*
> ```objc
>   std::optional<u16> GetRevision(const Partition& partition = PARTITION_NONE) const override
>   {
>     return 0;
> ```

Would it be better to just return {} (i.e. no value) instead?

---

*[Source/Core/DiscIO/VolumeWiiCrypted.cpp, line 252 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/5530#-Klo9cZZhUjS7Sx8r8Lx:-Klo9cZZhUjS7Sx8r8Ly:bwn9to8) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d28c253e74c37e4f506f7d5c551e7c21b24401b4/Source/Core/DiscIO/VolumeWiiCrypted.cpp#L252)):*
> ```cpp
> std::optional<u16> CVolumeWiiCrypted::GetRevision(const Partition& partition) const
> {
>   return ReadSwapped<u16>(7, partition);
> ```

The old code seems to read only one byte, which is then zero extended to a u16. With this change, we're reading 2 from the volume. Is this intended?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/5530)*

**Comment by BhaaLseN** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/Core/Core/ConfigManager.cpp, line 723 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/5530#-Klo9-8MRf9NysAKSkns:-KloF4IrAI-K5ULs0WHI:bdifa4b) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d28c253e74c37e4f506f7d5c551e7c21b24401b4/Source/Core/Core/ConfigManager.cpp#L723)):*
<details><summary><i>Previously, leoetlino (Leo Lam) wrote…</i></summary><blockquote>

Would it be better to just use value_or(0) instead of requiring a valid revision?
</blockquote></details>

Same problem as with title id if we skip the call?

---

*[Source/Core/Core/Boot/Boot_BS2Emu.cpp, line 95 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/5530#-KloEh2e_rBoPKHBUW1D:-KloEh2f6fg7nHDjHAmi:b-u7gla3) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d28c253e74c37e4f506f7d5c551e7c21b24401b4/Source/Core/Core/Boot/Boot_BS2Emu.cpp#L95)):*
> ```cpp
>   const std::optional<u32> size = volume.ReadSwapped<u32>(offset + 0x14, partition);
>   const std::optional<u32> trailer = volume.ReadSwapped<u32>(offset + 0x18, partition);
>   if (!entry || !size || !trailer || *entry == (u32)-1 || *size + *trailer == (u32)-1)
> ```

What about `entry.or_value(-1) == -1` etc. instead? Will be interresting for `size + trailer` though.

---

*[Source/Core/DiscIO/VolumeWiiCrypted.cpp, line 252 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/5530#-Klo9cZZhUjS7Sx8r8Lx:-KloG1HKhyplPFcp86Fj:b-47naq) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d28c253e74c37e4f506f7d5c551e7c21b24401b4/Source/Core/DiscIO/VolumeWiiCrypted.cpp#L252)):*
<details><summary><i>Previously, leoetlino (Leo Lam) wrote…</i></summary><blockquote>

The old code seems to read only one byte, which is then zero extended to a u16. With this change, we're reading 2 from the volume. Is this intended?
</blockquote></details>

Same in `CVolumeGC::GetRevision`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/5530)*

**Comment by JosJuice** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/Core/Core/ConfigManager.cpp, line 723 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/5530#-Klo9-8MRf9NysAKSkns:-KloJGddz7M7IrTXRsea:bdnrfg5) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d28c253e74c37e4f506f7d5c551e7c21b24401b4/Source/Core/Core/ConfigManager.cpp#L723)):*
<details><summary><i>Previously, BhaaLseN (BhaaL) wrote…</i></summary><blockquote>

Same problem as with title id if we skip the call?
</blockquote></details>

@leoetlino Yeah, that would be better. Let's use whatever data we can get our hands on even if we can't get it all.

@BhaaLseN Not really. `GetRevision` only fails if something is super wrong (or if we're using VolumeDirectory after Léo's suggestion of making it return `{}`, but I'm fine with counting VolumeDirectory as something really wrong).

---

*[Source/Core/Core/Boot/Boot_BS2Emu.cpp, line 95 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/5530#-KloEh2e_rBoPKHBUW1D:-KloJwOGL-bpzh7XZRPO:b7ak5ol) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d28c253e74c37e4f506f7d5c551e7c21b24401b4/Source/Core/Core/Boot/Boot_BS2Emu.cpp#L95)):*
<details><summary><i>Previously, BhaaLseN (BhaaL) wrote…</i></summary><blockquote>

What about `entry.or_value(-1) == -1` etc. instead? Will be interresting for `size + trailer` though.
</blockquote></details>

It would probably just end up looking confusing, especially since we only can do it for `entry`.

---

*[Source/Core/DiscIO/VolumeDirectory.h, line 54 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/5530#-Klo8u90c3-ANZHawGMM:-KloK6_KAZIWKzmUd4Ej:b-7r76f6) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d28c253e74c37e4f506f7d5c551e7c21b24401b4/Source/Core/DiscIO/VolumeDirectory.h#L54)):*
<details><summary><i>Previously, leoetlino (Leo Lam) wrote…</i></summary><blockquote>

Would it be better to just return {} (i.e. no value) instead?
</blockquote></details>

Sure, why not. (But VolumeDirectory's life will probably come to an end soon anyway...)

---

*[Source/Core/DiscIO/VolumeWiiCrypted.cpp, line 252 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/5530#-Klo9cZZhUjS7Sx8r8Lx:-KloKQ_dr7pIrTnvY3EV:b-dtr4ld) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d28c253e74c37e4f506f7d5c551e7c21b24401b4/Source/Core/DiscIO/VolumeWiiCrypted.cpp#L252)):*
<details><summary><i>Previously, BhaaLseN (BhaaL) wrote…</i></summary><blockquote>

Same in `CVolumeGC::GetRevision`
</blockquote></details>

Thanks for catching that. I've fixed it in both places.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/5530)*

**Comment by leoetlino** [comment]
:lgtm:

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/5530#-:-KlpH_vQ9a9_LEKiWCg4:bnfp4nl)*

**Comment by leoetlino** [comment]
This now needs a rebase.

**Pull request merged**